mm2src/coins_activation/src/eth_with_token_activation.rs

```diff
@@ -0,0 +1,263 @@
+    fn from(err: EthActivationV2Error) -> Self {
+        match err {
+            EthActivationV2Error::InvalidPayload(e) => EnablePlatformCoinWithTokensError::InvalidPayload(e),
+            EthActivationV2Error::ActivationFailed { ticker, error } => {
+                EnablePlatformCoinWithTokensError::PlatformCoinCreationError { ticker, error }
+            },
+            EthActivationV2Error::CouldNotFetchBalance(e)
+            | EthActivationV2Error::UnreachableNodes(e)
+            | EthActivationV2Error::AtLeastOneNodeRequired(e) => EnablePlatformCoinWithTokensError::Transport(e),
+            EthActivationV2Error::InternalError(e) => EnablePlatformCoinWithTokensError::Internal(e),
+        }
+    }
+}
+
+impl TryFromCoinProtocol for CoinProtocol {
+    fn try_from_coin_protocol(proto: CoinProtocol) -> Result<Self, MmError<CoinProtocol>>
+    where
+        Self: Sized,
+    {
+        Ok(proto)
+    }
+}
+
+pub struct Erc20Initializer {
+    platform_coin: EthCoin,
+}
+
+impl From<EthActivationV2Error> for InitTokensAsMmCoinsError {
+    fn from(error: EthActivationV2Error) -> Self {
+        match error {
+            EthActivationV2Error::InvalidPayload(e) => InitTokensAsMmCoinsError::InvalidPayload(e),
+            EthActivationV2Error::ActivationFailed { ticker, error } => {
+                InitTokensAsMmCoinsError::InitializationFailed { ticker, error }
+            },
+            EthActivationV2Error::CouldNotFetchBalance(e)
+            | EthActivationV2Error::UnreachableNodes(e)
+            | EthActivationV2Error::AtLeastOneNodeRequired(e)
+            | EthActivationV2Error::InternalError(e) => InitTokensAsMmCoinsError::Internal(e),
+        }
+    }
+}
+
+#[async_trait]
+impl TokenInitializer for Erc20Initializer {
+    type Token = EthCoin;
+    type TokenActivationRequest = EthActivationV2Request;
+    type TokenProtocol = CoinProtocol;
+    type InitTokensError = EthActivationV2Error;
+
+    fn tokens_requests_from_platform_request(
+        platform_params: &EthWithTokensActivationRequest,
+    ) -> Vec<TokenActivationRequest<Self::TokenActivationRequest>> {
+        platform_params.erc20_tokens_requests.clone()
+    }
+
+    async fn enable_tokens(
+        &self,
+        activation_params: Vec<TokenActivationParams<EthActivationV2Request, CoinProtocol>>,
+    ) -> Result<Vec<EthCoin>, MmError<EthActivationV2Error>> {
+        let ctx = MmArc::from_weak(&self.platform_coin.ctx())
+            .ok_or("No context")
+            .map_err(|e| EthActivationV2Error::InternalError(e.to_string()))?;
+
+        let secret = CryptoCtx::from_ctx(&ctx)
+            .map_err(|e| EthActivationV2Error::InternalError(e.to_string()))?
+            .iguana_ctx()
+            .secp256k1_privkey_bytes()
+            .to_vec();
+
+        let mut tokens = vec![];
+        for param in activation_params {
+            let coins_en = coin_conf(&ctx, &param.ticker);
+            let coin: EthCoin = eth_coin_from_conf_and_request_v2(
```

Comment:
`eth_coin_from_conf_and_request_v2` is not required here. `enable_erc20` should be as simple as `enable_slp`: it can copy/clone almost all fields from platform coin, skipping additional initialization steps. For example, `eth_coin_from_conf_and_request_v2` checks Ethereum client version for each node. If you activate ETH + 30 tokens, there will be 31 `web3_clientVersion` per each URL, causing excessive traffic usage and slowing the activation down.

Reply:
done